Plutarch/Verify.hs

```diff
@@ -0,0 +1,277 @@
+    ver <- tcont $ plet $ ppairDataBuiltin # fst' # snd'
+    pure $ (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( PTryFrom PData (PAsData b)
+  , PTryFrom PData (PAsData (PDataRecord xs))
+  , x ~ (s ':= b)
+  , PIsData (PDataRecord xs)
+  , PExcess PData (PAsData (PDataRecord xs)) ~ PDataRecord xs
+  ) =>
+  PTryFrom PData (PAsData (PDataRecord (x ': xs)))
+  where
+  type PExcess PData (PAsData (PDataRecord (x ': xs))) = PDataRecord (x ': xs)
+  ptryFrom opq = do
+    lst <- tcont $ plet (pfromData @(PBuiltinList _) $ punsafeCoerce opq)
+    let lsthead :: Term _ PData
+        lsthead = phead # lst
+        lsttail :: Term _ (PAsData (PBuiltinList PData))
+        lsttail = pdata $ ptail # lst
+        verhead :: Term _ (PAsData b)
+        verhead = unTermCont $ fst <$> (ptryFrom @PData @(PAsData b) lsthead)
+        vertail :: Term _ (PDataRecord xs)
+        vertail = unTermCont $ snd <$> (ptryFrom @PData @(PAsData (PDataRecord xs)) (pforgetData lsttail))
+    ver <-
+      tcont $
+        plet $
+          pdcons @s
+            # verhead
+            # vertail
+    pure (punsafeCoerce opq, ver)
+
+instance
+  {-# OVERLAPPING #-}
+  ( PTryFrom PData (PAsData b)
+  , PIsData b
+  , x ~ (s ':= b)
+  ) =>
+  PTryFrom PData (PAsData (PDataRecord '[x]))
+  where
+  type PExcess PData (PAsData (PDataRecord '[x])) = PDataRecord '[x]
+  ptryFrom opq = do
+    let lsthead :: Term _ PData
+        lsthead = phead # (pfromData @(PBuiltinList _) $ punsafeCoerce opq)
+        verhead :: Term _ (PAsData b)
+        verhead = unTermCont $ fst <$> (ptryFrom @PData @(PAsData b) lsthead)
+    ver <-
+      tcont $
+        plet $
+          pdcons @s
+            # verhead
+            # pdnil
+    pure $ (punsafeCoerce opq, ver)
+
+----------------------- POpaque Instances -----------------------------------------------
+
+{- |
+    for none of the opaque instances it can be verified
+    that the actual structure is what it says to be
+    because that data is lost when the PAsData wrapper
+    is removed, this can only be safely used if you obtained
+    your POpaque safely
+-}
+instance
+  ( PTryFrom PData (PAsData a)
+  , PIsData a
+  ) =>
+  PTryFrom POpaque a
+  where
+  type PExcess POpaque a = PAsData a
+  ptryFrom opq = do
+    let prop :: Term _ a
+        prop = punsafeCoerce opq
+    ver' <- fst <$> (ptryFrom @PData @(PAsData a) $ pforgetData $ pdata prop)
+    ver <- tcont $ plet ver'
+    pure $ (punsafeCoerce opq, ver)
+
+----------------------- Helper functions ------------------------------------------------
+
+pchooseData :: Term s (PData :--> a :--> a :--> a :--> a :--> a :--> a)
+pchooseData = phoistAcyclic $ pforce $ punsafeBuiltin PLC.ChooseData
+
+pcheckType :: (Term s PInteger) -> Term _ (PData :--> PAsData b)
+pcheckType i = plam $ \d ->
+  let con :: Term _ PInteger
+      con = pchooseData # d # 0 # 1 # 2 # 3 # 4
+   in pif (con #== i) (punsafeCoerce d) perror
```

Comment:
this is essentially equivalent to the `UnConstrData`/`UnListData`/`UnMapData`/`UnIData`/`UnBData` functions since it `perror`s anyway. `ChooseData` is really only relevant when you _want_ a recovery path.

Reply:
that's why I exported the functions for future use in a recovering version of this. Perhaps I should refactor it out to another module tho?

Reply:
You can leave them out of this PR for now. I suspect there's a better way but we don't need this right now.

Reply:
copied.